internal/sync/waitgroup.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
+// SPDX-License-Identifier: MIT
+
+// Package sync extends basic synchronization primitives.
+package sync
+
+import (
+	"sync"
+)
+
+// A WaitGroup waits for a collection of goroutines to finish.
+// The main goroutine calls Add to set the number of
+// goroutines to wait for. Then each of the goroutines
+// runs and calls Done when finished. At the same time,
+// Wait can be used to block until all goroutines have finished.
+//
+// WaitGroups in the sync package do not allow adding or
+// subtracting from the counter while another goroutine is
+// waiting, while this one does.
```

Comment:
I was looking at https://pkg.go.dev/sync#WaitGroup.Add and I don't believe this is strictly accurate (I know you are just migrating this package and didn't initially introduce it). Per https://pkg.go.dev/sync#WaitGroup.Add,
```
Note that calls with a positive delta that occur when the counter is zero must happen before a Wait. Calls with a negative delta, or calls with a positive delta that start when the counter is greater than zero, may happen at any time. Typically this means the calls to Add should execute before the statement creating the goroutine or other event to be waited for.
```

I believe the problem that was being encountered in the initial race condition was the rare case in which the wait group counter went to zero, but then [`Accept()` was called on the listener](https://github.com/pion/udp/blob/d4aece4f6a32af411ee19e87f7051e75dd95ab16/conn.go#L48), and its `Add()` raced with `Wait() in the goroutine that is [started when listening beings](https://github.com/pion/udp/blob/d4aece4f6a32af411ee19e87f7051e75dd95ab16/conn.go#L151).

I think the longer term solution is to consider some slight redesign in the UDP listener, but in the mean time, perhaps we could update this comment to mention that it guards against race conditions when adding to a wait group with counter 0 that is concurrently being waited on. What do you think?

Reply:
Thanks for investigation. I was also sceptical about this comment, but then didnt go into the details as I was not the original author.

Looking again at this, I dont believe it would be a good decission to merge this as is.
In my opinion the race should be fixed in the listener not by patching a commonly understood synchronization primitive.

Reply:
@stv0g sounds good to me -- are you planning to address in this PR or another?

Reply:
Hey @hasheddan,

Sorry, I am currently too busy to look into this in detail.

Would you have the capacity to take this over?
Or should we go with the old fix instead?

Reply:
Hey @stv0g! I can take a look, thanks for the ping!